components/style/properties.mako.rs

```diff
@@ -6798,56 +6655,92 @@ pub fn cascade(viewport_size: Size2D<Au>,
+                                                              &mut error_reporter);
                 }
-                let discriminant = unsafe {
-                    intrinsics::discriminant_value(declaration) as usize
-                };
-                (cascade_property[discriminant].unwrap())(declaration,
-                                                          &mut style,
-                                                          inherited_style,
-                                                          &context,
-                                                          &mut seen,
-                                                          &mut cacheable,
-                                                          &mut error_reporter);
-            }
-        }
-    });
-
-    // The initial value of border-*-width may be changed at computed value time.
-    {
-        let border = Arc::make_mut(&mut style.border);
-        % for side in ["top", "right", "bottom", "left"]:
-            // Like calling to_computed_value, which wouldn't type check.
-            if !context.border_${side}_present {
-                border.border_${side}_width = Au(0);
             }
         % endfor
-    }
+    });
 
     // The initial value of display may be changed at computed value time.
     if !seen.get_display() {
-        let box_ = Arc::make_mut(&mut style.box_);
-        let computed_value = box_.display.to_computed_value(&context);
-        box_.display = computed_value;
-        box_._servo_display_for_hypothetical_box =
+        let computed_value = context.style.box_.display.to_computed_value(&context);
+        Arc::make_mut(&mut context.style.box_).display = computed_value;
+        Arc::make_mut(&mut context.style.box_)._servo_display_for_hypothetical_box =
             longhands::_servo_display_for_hypothetical_box::derive_from_display(
                     computed_value,
                     &context);
     }
 
+    let mut style = context.style;
+
+    {
+        use computed_values::overflow_x::T as overflow;
+        use computed_values::overflow_y;
+        match (style.box_.overflow_x, style.box_.overflow_y.0) {
+            (overflow::visible, overflow::visible) => {}
+            (overflow::visible, _) => {
+                Arc::make_mut(&mut style.box_).overflow_x = overflow::auto
+            }
+            (_, overflow::visible) => {
+                Arc::make_mut(&mut style.box_).overflow_y = overflow_y::T(overflow::auto)
+            }
+            _ => {}
+        }
+    }
+
+    // The initial value of border-*-width may be changed at computed value time.
+    % for side in ["top", "right", "bottom", "left"]:
+        // Like calling to_computed_value, which wouldn't type check.
+        if style.border.border_${side}_style.none_or_hidden() &&
+           style.border.border_${side}_width != Au(0) {
+            Arc::make_mut(&mut style.border).border_${side}_width = Au(0);
```

Comment:
Pushing the `Arc::make_mut()` into the if condition like you did here may improve cascading performance significantly, FWIW. Thanks!